File: generated-code/codecraft/csharp/Model/BuildProperties.cs
namespace TransGenTest.Model
{
    public struct BuildProperties
    {
        public Model.EntityType[] Options { get; set; }
        public int? InitHealth { get; set; }
    
        public BuildProperties(Model.EntityType[] options, int? initHealth)
        {
            this.Options = options;
            this.InitHealth = initHealth;
        }
    
        public static BuildProperties ReadFrom(System.IO.BinaryReader reader)
        {
            var result = new BuildProperties();
            result.Options = new Model.EntityType[reader.ReadInt32()];
            for (int optionsIndex = 0; optionsIndex < result.Options.Length; optionsIndex++)
            {
                result.Options[optionsIndex] = EntityTypeHelper.ReadFrom(reader);
            }
            if (reader.ReadBoolean())
            {
                result.InitHealth = reader.ReadInt32();
            } else
            {
                result.InitHealth = null;
            }
            return result;
        }
    
        public void WriteTo(System.IO.BinaryWriter writer)
        {
            writer.Write(Options.Length);
            foreach (var optionsElement in Options)
            {
                writer.Write((int) (optionsElement));
            }
            if (!InitHealth.HasValue)
            {
                writer.Write(false);
            } else
            {
                writer.Write(true);
                writer.Write(InitHealth.Value);
            }
        }
    
        public override string ToString() {
            string stringResult = "BuildProperties { ";
            stringResult += "Options: ";
            stringResult += "[ ";
            int optionsIndex = 0;
            foreach (var optionsElement in Options)
            {
                if (optionsIndex != 0) {
                    stringResult += ", ";
                }
                stringResult += optionsElement.ToString();
                optionsIndex++;
            }
            stringResult += " ]";
            stringResult += ", ";
            stringResult += "InitHealth: ";
            if (!InitHealth.HasValue)
            {
                stringResult += "null";
            } else
            {
                stringResult += InitHealth.Value.ToString();
            }
            stringResult += " }";
            return stringResult;
        }
    }
}

File: generated-code/codecraft/fsharp/Model/BuildProperties.fs
#nowarn "0058"

namespace TransGenTest.Model

/// Entity's build properties
type BuildProperties = {
    /// Valid new entity types
    Options: EntityType[];
    /// Initial health of new entity. If absent, it will have full health
    InitHealth: option<int>;
} with

    /// Write BuildProperties to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write this.Options.Length
        this.Options |> Array.iter (fun value ->
            writer.Write (int value) )
        match this.InitHealth with
            | Some value ->
                writer.Write true
                writer.Write value
            | None -> writer.Write false

    /// Read BuildProperties from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        Options = [|for _ in 1 .. reader.ReadInt32() do
                      yield reader.ReadInt32() |> enum |]
        InitHealth = match reader.ReadBoolean() with
                         | true -> Some(reader.ReadInt32())
                         | false -> None
    }